File: api/src/services/mdblist.rs
use std::sync::Arc;

use crate::error::AppError;
use crate::id::MediaType;
use crate::services::api_key_pool::ApiKeyPool;
use crate::services::retry::{self, MDBLIST_RETRY};
use serde::Deserialize;

#[derive(Clone)]
pub struct MdblistClient {
    key_pool: Arc<ApiKeyPool>,
    http: reqwest::Client,
}

#[derive(Debug, Deserialize)]
pub struct MdblistResponse {
    #[serde(default)]
    pub ratings: Vec<MdblistRating>,
    #[serde(default)]
    pub ids: MdblistIds,
    #[serde(default)]
    pub score: Option<f64>,
}

#[derive(Debug, Default, Deserialize)]
pub struct MdblistIds {
    pub imdb: Option<String>,
    pub tmdb: Option<u64>,
    pub tvdb: Option<u64>,
}

#[derive(Debug, Deserialize)]
pub struct MdblistRating {
    pub source: String,
    pub value: Option<f64>,
    pub score: Option<f64>,
    pub votes: Option<i64>,
}

fn mdblist_kind(media_type: &MediaType) -> Result<&'static str, AppError> {
    match media_type {
        MediaType::Movie => Ok("movie"),
        MediaType::Tv => Ok("show"),
        MediaType::Episode => Err(AppError::Other("mdblist does not support episode ratings".into())),
    }
}

fn imdb_ratings_url(kind: &str, imdb_id: &str) -> String {
    format!("https://api.mdblist.com/imdb/{kind}/{imdb_id}")
}

fn tmdb_ratings_url(kind: &str, tmdb_id: u64) -> String {
    format!("https://api.mdblist.com/tmdb/{kind}/{tmdb_id}")
}

impl MdblistClient {
    pub fn new(key_pool: ApiKeyPool, http: reqwest::Client) -> Self {
        Self { key_pool: Arc::new(key_pool), http }
    }

    pub fn active_key_hash(&self) -> String {
        self.key_pool.active_key_hash()
    }

    async fn fetch(&self, url: &str) -> Result<MdblistResponse, AppError> {
        let api_key = self.key_pool.active_key_raw();
        let key_hash = self.key_pool.active_key_hash();

        let url = url.to_owned();
        let resp = retry::send_with_retry(&MDBLIST_RETRY, || {
            self.http
                .get(&url)
                .query(&[("apikey", api_key.as_str())])
                .send()
        })
        .await?;

        let was_429 = resp.status() == reqwest::StatusCode::TOO_MANY_REQUESTS;

        let result = resp.error_for_status();

        if was_429 {
            self.key_pool.report_429();
            tracing::warn!(
                key = %key_hash,
                "mdblist API key rate limited (429)"
            );
        } else {
            self.key_pool.report_success();
        }

        let resp = result?;
        Ok(resp.json().await?)
    }

    pub async fn get_ratings(
        &self,
        imdb_id: &str,
        media_type: &MediaType,
    ) -> Result<MdblistResponse, AppError> {
        let kind = mdblist_kind(media_type)?;
        self.fetch(&imdb_ratings_url(kind, imdb_id)).await
    }

    pub async fn get_ratings_by_tmdb(
        &self,
        tmdb_id: u64,
        media_type: &MediaType,
    ) -> Result<MdblistResponse, AppError> {
        let kind = mdblist_kind(media_type)?;
        self.fetch(&tmdb_ratings_url(kind, tmdb_id)).await
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mdblist_kind_maps_movie_and_tv() {
        assert_eq!(mdblist_kind(&MediaType::Movie).unwrap(), "movie");
        assert_eq!(mdblist_kind(&MediaType::Tv).unwrap(), "show");
    }

    #[test]
    fn mdblist_kind_rejects_episode() {
        assert!(mdblist_kind(&MediaType::Episode).is_err());
    }

    #[test]
    fn imdb_ratings_url_format() {
        assert_eq!(imdb_ratings_url("show", "tt2560140"), "https://api.mdblist.com/imdb/show/tt2560140");
        assert_eq!(imdb_ratings_url("movie", "tt0111161"), "https://api.mdblist.com/imdb/movie/tt0111161");
    }

    #[test]
    fn tmdb_ratings_url_format() {
        assert_eq!(tmdb_ratings_url("show", 1429), "https://api.mdblist.com/tmdb/show/1429");
        assert_eq!(tmdb_ratings_url("movie", 550), "https://api.mdblist.com/tmdb/movie/550");
    }
}

File: api-go/internal/services/mdblist.go
package services

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"openposterdb/internal/errors"

	"log/slog"
)

type MdblistClient struct {
	KeyPool *APIKeyPool
	HTTP    *http.Client
}

func NewMdblistClient(keys []string, httpClient *http.Client) *MdblistClient {
	return &MdblistClient{
		KeyPool: NewAPIKeyPool(keys),
		HTTP:    httpClient,
	}
}

func (c *MdblistClient) ActiveKeyHash() string {
	return c.KeyPool.ActiveKeyHash()
}

type MdblistResponse struct {
	Ratings []MdblistRating `json:"ratings"`
	IDs     MdblistIDs      `json:"ids"`
	Score   *float64        `json:"score"`
}

type MdblistIDs struct {
	IMDB *string `json:"imdb"`
	TMDB *uint64 `json:"tmdb"`
	TVDB *uint64 `json:"tvdb"`
}

type MdblistRating struct {
	Source string   `json:"source"`
	Value  *float64 `json:"value"`
	Score  *float64 `json:"score"`
	Votes  *int64   `json:"votes"`
}

func (c *MdblistClient) GetRatingsCtx(ctx context.Context, imdbID, mediaType string) (*MdblistResponse, error) {
	kind, err := mdblistKind(mediaType)
	if err != nil {
		return nil, err
	}
	url := fmt.Sprintf("https://api.mdblist.com/imdb/%s/%s", kind, imdbID)
	return c.fetchCtx(ctx, url)
}

func (c *MdblistClient) GetRatings(imdbID, mediaType string) (*MdblistResponse, error) {
	return c.GetRatingsCtx(context.Background(), imdbID, mediaType)
}

func (c *MdblistClient) GetRatingsByTMDBCtx(ctx context.Context, tmdbID uint64, mediaType string) (*MdblistResponse, error) {
	kind, err := mdblistKind(mediaType)
	if err != nil {
		return nil, err
	}
	url := fmt.Sprintf("https://api.mdblist.com/tmdb/%s/%d", kind, tmdbID)
	return c.fetchCtx(ctx, url)
}

func (c *MdblistClient) GetRatingsByTMDB(tmdbID uint64, mediaType string) (*MdblistResponse, error) {
	return c.GetRatingsByTMDBCtx(context.Background(), tmdbID, mediaType)
}

func (c *MdblistClient) fetchCtx(ctx context.Context, url string) (*MdblistResponse, error) {
	apiKey := c.KeyPool.ActiveKeyRaw()
	keyHash := c.KeyPool.ActiveKeyHash()

	fullURL := fmt.Sprintf("%s?apikey=%s", url, apiKey)

	start := time.Now()
	resp, err := SendWithRetry(&MDBListRetry, func() (*http.Response, error) {
		req, _ := http.NewRequestWithContext(ctx, "GET", fullURL, nil)
		req.Header.Set("User-Agent", "openposterdb/1.2.1")
		return c.HTTP.Do(req)
	})
	logSlow("MDBList", time.Since(start).Milliseconds())

	if err != nil {
		return nil, errors.NewAPIError(err)
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	if resp.StatusCode == http.StatusTooManyRequests {
		c.KeyPool.Report429()
		slog.Warn("mdblist API key rate limited (429)", "key", keyHash)
		return nil, errors.NewOther(fmt.Sprintf("MDBList 429: %s", string(body)))
	}

	c.KeyPool.ReportSuccess()

	if resp.StatusCode != http.StatusOK {
		return nil, errors.NewOther(fmt.Sprintf("MDBList returned %d: %s", resp.StatusCode, string(body)))
	}

	var result MdblistResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, errors.NewAPIError(err)
	}
	return &result, nil
}

func (c *MdblistClient) fetch(url string) (*MdblistResponse, error) {
	return c.fetchCtx(context.Background(), url)
}

func mdblistKind(mediaType string) (string, error) {
	switch mediaType {
	case "movie":
		return "movie", nil
	case "tv":
		return "show", nil
	default:
		return "", errors.NewOther("mdblist does not support episode ratings")
	}
}
